src/simplify_reshapes.cpp

```diff
@@ -949,6 +981,76 @@ struct find_reshape_reshape_dot
     }
 };
 
+struct find_reshape_const_dot
+{
+    auto matcher() const
+    {
+        return match::name("dot")(
+            match::used_once(),
+            match::either_arg(0, 1)(match::name("reshape").bind("rsp"),
+                                    match::skip_broadcasts(match::is_constant().bind("constant"))));
+    }
+
+    void apply(module& m, const match::matcher_result& r) const
+    {
+        auto dot      = r.result;
+        auto rsp      = r.instructions["rsp"];
+        auto constant = r.instructions["constant"];
+
+        auto const_lens = constant->get_shape().lens();
+        if(const_lens.size() > 2)
+            return;
+
+        auto rsp_lens = rsp->get_shape().lens();
+        auto inp      = rsp->inputs().front();
+        auto inp_lens = inp->get_shape().lens();
+
+        // Gemm axis should not be altered by the reshape
+        bool flipped    = rsp == dot->inputs().back();
+        size_t dot_axis = (flipped) ? -2 : -1;
+        if(rsp_lens.end()[dot_axis] != inp_lens.end()[dot_axis])
```

Comment:
I don't get what this doing. `end()[dot_axis]`  What does is it  doing ?

Reply:
The goal here is to check that the gemm axis is unaffected by the reshape. So it A is the input and W is the weight, then for A*W, the last axis of A shouldnt be manipulated by the reshape, and if its W*A, then the second last axis should be untouched.

Reply:
I just didn't understand syntax. That's all.  
Also this should do some bound checking e.g. what is `end()[dot_axis]` can go out of bound for some weird case.

Reply:
rsp_lens cant be out of bounds because dot op operands need to have dimns >= 2, but the input to that can definitely be and issue so added a check for that.